typing/parmatch.ml

```diff
@@ -1869,14 +1876,28 @@ let extendable_path path =
     Path.same path Predef.path_unit ||
     Path.same path Predef.path_option)
 
+let warn_fragile_type ty env =
+  let _,_,decl = Ctype.extract_concrete_typedecl env ty in
+  let warn = Warnings.Fragile_match "" in
+  match Warnings.status warn with
+    | Warnings.Always -> true
+    | Warnings.Never -> false
+    | Warnings.Implicit | Warnings.Explicit ->
+        Builtin_attributes.with_warning_attribute
+          decl.Types.type_attributes
+          (fun () -> Warnings.(is_active warn))
+
 let rec collect_paths_from_pat r p = match p.pat_desc with
 | Tpat_construct(_, {cstr_tag=(Cstr_constant _|Cstr_block _|Cstr_unboxed)},ps)
   ->
-    let path =  get_type_path p.pat_type p.pat_env in
-    List.fold_left
-      collect_paths_from_pat
-      (if extendable_path path then add_path path r else r)
-      ps
+    (match get_type_path p.pat_type p.pat_env with
+     | Some path ->
+         List.fold_left
+           collect_paths_from_pat
+           (if extendable_path path && warn_fragile_type p.pat_type p.pat_env
+            then add_path path r else r)
+           ps
+     | None -> r)
```

Comment:
@trefis you're right. The `None` case is here to avoid a fatal error in a particular corner case (type error within a recursive module, as described in [Mantis issue 6394](https://caml.inria.fr/mantis/view.php?id=6394)). A proper type error will be generated somewhat later, and we can mostly ignore the patterns related to that type.

Regarding the way the code is written, your guess is wrong 😄 . It's just that the modification done for supporting per-type warn 4 and fixing PR 6394 were done in two separate commits and I forgot to refactor the code at that time. Your proposal is indeed clearer, and I'll fix up the code.